content/blog/2017/2017-02-14-declarative-notifications.adoc

```diff
@@ -0,0 +1,436 @@
+      to: 'bitwiseman@bitwiseman.com',
+      subject: subject,
+      body: details,
+      recipientProviders: [[$class: 'DevelopersRecipientProvider']]
+    )
+}
+----
+
+image::/images/post-images/2017-02-14/blueocean-notifications.png[Global Pipeline Library, role="center"]
+
+== Move Notifications to Shared Library
+
+Our current pipeline is pretty good, we have our Declarative Pipeline sending notifications.
+However, the notification method taking up about a third of our `Jenkinsfile` is a bit of a distraction.
+Also, as Jenkins admin, I'd rather share that notification method among all my projects
+without copying it to each new project and then having maintain all the copies.
+I want to move that method to a Shared Library.
+
+Shared Libraries are not specific to Declarative. They were released in their
+current form several months ago and were functional and useful in Scripted Pipeline.
+Not every method like this one should go into a Shared Library,
+but Declarative's focus on overall pipeline flow makes it much clearer what
+parts of a pipeline are candidates for moving to a shared library.
+
+
+=== Setup a Shared Library
+
+I've created a simple shared library repository for this series of posts, called
+link:https://github.com/bitwiseman/jenkins-pipeline-shared[bitwiseman/jenkins-pipeline-shared].
+The Shared Library functionality has too many configuration options to cover in one post.
+I've chosen to set mine up a "Global" Library, accessible from any project on my Jenkin master.
+Under "Manage Jenkins" -> "Configure System" -> "Global Pipeline Libraries",
+I've chosen the name `bitwiseman-shared` and pointed it at my repository.
+I've set the default branch to for the library to `master`, but I'll override that in my pipeline.
+
+image::/images/post-images/2017-02-14/shared-library.png[Global Pipeline Library, role="center"]
+
+=== Move the code to the Library
+
+Moving a method to a Shared Library involves creating a file with the name of our
+method, adding our method to that file as a `call()` method, adding an optional
+help text file to match, and seeing if we need to add a `@Library` directive to
+our `Jenkinsfile`.
+
+There are few limitation for methods in Shared Libraries, but they don't apply here.
+The minimal set of dependencies for `sendNotifications` means I can once again
+basically copy and paste the code across and add short bit of help text for completeness.
+I'll check this change into a branch name in the library,
+`blog/declarative/notifications` the same as my branch in the other repository.
+This will let me make changes on the master branch later without breaking this example.
+I'll use use the `@Library` directive to tell Jenkins to use that version with this Pipeline.
+
+.vars/sendNotifications.groovy
+[source, groovy]
+----
+#!/usr/bin/env groovy
+
+/**
+ * Send notifications based on build status string
+ */
+def call(String buildStatus = 'STARTED') {
+  // build status of null means successful
```

Comment:
This is the third time in this blog post that the full method body has been pasted. I suggest just linking to the full file on GitHub so we can get to the `Jenkinsfile` faster here.


If you're married to including all this text (again!) then move the `Jenkinsfile` to the top.